File: zad_10.py
from math import *

def simpson_rule(func, a, b, nseg):
    if nseg % 2 == 1:
        nseg += 1
    dx = (b - a) / nseg
    total = func(a) + func(b)
    for i in range(1, nseg, 2):
        total += 4 * func(a + i * dx)
    for i in range(2, nseg, 2):
        total += 2 * func(a + i * dx)
    return total * dx / 3

def func(x):
    return 1 / (x**2 - 4*x + 3)

print("Using Simpson's method")
print("Integrating function: f(x) = 1 / (x^2 - 4x + 3)")

# Avoid the singularity at x=1 by splitting the interval
delta = 1e-3  # Wider margin to improve stability
a1, b1 = 0, 1 - delta
a2, b2 = 1 + delta, 2

eps = 1e-7

# Left side of the singularity
nseg = 2
int1_1 = simpson_rule(func, a1, b1, nseg)
int1_2 = simpson_rule(func, a1, b1, nseg * 2)

while abs(int1_1 - int1_2) > eps:
    nseg *= 2
    int1_1 = simpson_rule(func, a1, b1, nseg)
    int1_2 = simpson_rule(func, a1, b1, nseg * 2)

int_left = int1_2

# Right side of the singularity
nseg = 2
int2_1 = simpson_rule(func, a2, b2, nseg)
int2_2 = simpson_rule(func, a2, b2, nseg * 2)

while abs(int2_1 - int2_2) > eps:
    nseg *= 2
    int2_1 = simpson_rule(func, a2, b2, nseg)
    int2_2 = simpson_rule(func, a2, b2, nseg * 2)

int_right = int2_2

print("\nAnswer: I =", int_left + int_right)
print("Number of segments:", nseg * 2)

File: zad_10.cpp
#include <iostream>
#include <cmath>
using namespace std;

double func(double x) {
    return 1.0 / (x * x - 4 * x + 3);
}

double simpson(double (*f)(double), double a, double b, int nseg) {
    if (nseg % 2 == 1) nseg++;
    double h = (b - a) / nseg;
    double result = f(a) + f(b);
    for (int i = 1; i < nseg; i += 2)
        result += 4 * f(a + i * h);
    for (int i = 2; i < nseg; i += 2)
        result += 2 * f(a + i * h);
    return result * h / 3.0;
}

int main() {
    double eps = 1e-7;
    double delta = 1e-3;
    double a1 = 0, b1 = 1 - delta;
    double a2 = 1 + delta, b2 = 2;

    int nseg = 2;
    double int1 = simpson(func, a1, b1, nseg);
    double int2 = simpson(func, a1, b1, nseg * 2);

    while (fabs(int1 - int2) > eps) {
        nseg *= 2;
        int1 = simpson(func, a1, b1, nseg);
        int2 = simpson(func, a1, b1, nseg * 2);
    }

    double left = int2;

    nseg = 2;
    int1 = simpson(func, a2, b2, nseg);
    int2 = simpson(func, a2, b2, nseg * 2);

    while (fabs(int1 - int2) > eps) {
        nseg *= 2;
        int1 = simpson(func, a2, b2, nseg);
        int2 = simpson(func, a2, b2, nseg * 2);
    }

    double right = int2;

    cout << "Integral 10 Result: " << left + right << endl;
    cout << "Segments used: " << nseg * 2 << endl;
    return 0;
}